### PR TITLE
On demand nodegroup backup

### LINKED ISSUE #629: Add On-Demand nodegroup as a backup to Spot
Create a spot nodegroup with a high priority and an on demand nodegroup with a low priority
- [prioritization between nodegroups](https://github.com/kubernetes/autoscaler/tree/589cd2af9c25088ce3408eeb88ba9239201d005d/cluster-autoscaler/expander/priority)
- [maximum total nodes across nodegroups](https://github.com/kubernetes/autoscaler/blob/589cd2af9c25088ce3408eeb88ba9239201d005d/cluster-autoscaler/main.go#L121)

Some logic like this would be nice:

```
if spin_up:
  if spot:
    try:
      spin_up_spot()
    catch:
      spin_up_on_demand()
  else:
    spin_up_on_demand()
if spin_down:
  spin_down_on_demand()
```

### PR DESCRIPTION
**Pull request by vishalbollu**
closes #629 

---

checklist:

- [ ] run `make test` and `make lint`
- [ ] test end to end manually (i.e. build/push all images, restart operator, and run `cortex deploy --refresh`)
- [ ] update the examples
- [ ] update the docs and add any new files to `summary.md`
- [ ] merge to master
- [ ] cherry-pick into release branches if applicable
- [ ] check that the docs are up-to-date in [gitbook](https://www.cortex.dev/v/master)
- [ ] alert the dev team if the dev environment changed
- [ ] delete the branch once it's merged

### CODE REVIEW

File: dev/versions.md [hunk 1]
```diff
@@ -151,8 +151,8 @@ Note: overriding horizontal-pod-autoscaler-sync-period on EKS is currently not s
...
-1. In the [GitHub Repo](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws), set the tree to the tag for the chosen release, and open `cloudprovider/aws/examples/cluster-autoscaler-autodiscover.yaml` (e.g. <https://github.com/kubernetes/autoscaler/blob/cluster-autoscaler-1.14.5/cluster-autoscaler/cloudprovider/aws/examples/cluster-autoscaler-autodiscover.yaml>)
+1. Find the latest patch release for our current version of k8s (e.g. k8s v1.14 -> cluster-autocluster v1.14.7) on [GitHub](https://github.com/kubernetes/autoscaler/releases) and check the changelog
+1. In the [GitHub Repo](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws), set the tree to the tag for the chosen release, and open `cloudprovider/aws/examples/cluster-autoscaler-autodiscover.yaml` (e.g. <https://github.com/kubernetes/autoscaler/blob/cluster-autoscaler-1.14.7/cluster-autoscaler/cloudprovider/aws/examples/cluster-autoscaler-autodiscover.yaml>)
 1. Copy the contents to `manager/manifests/cluster-autoscaler.yaml`
    1. Update this line of config:
```
Comment (deliahu) [inline dev/versions.md]: line 160 in this file should be updated

File: manager/manifests/cluster-autoscaler.yaml.j2 [hunk 2]
```diff
@@ -154,8 +164,12 @@ spec:
             - --stderrthreshold=info
             - --cloud-provider=aws
             - --skip-nodes-with-local-storage=false
-            - --expander=least-waste
```
Comment (deliahu) [inline manager/manifests/cluster-autoscaler.yaml.j2]: is least-waste the default?

File: pkg/lib/clusterconfig/clusterconfig.go [hunk 3]
```diff
@@ -850,6 +856,7 @@ func (cc *Config) UserFacingTable() table.KeyValuePairs {
 		items.Add(OnDemandPercentageAboveBaseCapacityUserFacingKey, *cc.SpotConfig.OnDemandPercentageAboveBaseCapacity)
 		items.Add(MaxPriceUserFacingKey, *cc.SpotConfig.MaxPrice)
 		items.Add(InstancePoolsUserFacingKey, *cc.SpotConfig.InstancePools)
+		items.Add(OnDemandBackupUserFacingKey, *cc.SpotConfig.InstancePools)
```
Comment (deliahu) [inline pkg/lib/clusterconfig/clusterconfig.go]: the value should be updated to the correct field

File: manager/render_template.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import yaml
+import os
+from copy import deepcopy
```
Comment (deliahu) [inline manager/render_template.py]: nit: deepcopy is not used

File: cli/cmd/lib_cluster_config.go [hunk 5]
```diff
@@ -302,7 +306,7 @@ func confirmInstallClusterConfig(clusterConfig *clusterconfig.Config, awsCreds *
 		fmt.Printf("this cluster will cost %s per hour if the minimum number of instances are running and %s per hour if the maximum number of instances are running%s\n\n", s.DollarsAndCents(totalMinPrice), s.DollarsAndCents(totalMaxPrice), spotSuffix)
 	}
 
-	if clusterConfig.Spot != nil && *clusterConfig.Spot {
+	if clusterConfig.Spot != nil && *clusterConfig.Spot && !clusterConfig.SpotConfig.OnDemandBackup {
```
[the remark this answers is not in the document]
Comment (vishalbollu) [reply cli/cmd/lib_cluster_config.go]: yes we do check it

**Pull request merged**